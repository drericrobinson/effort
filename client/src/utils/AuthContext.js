import React from 'react';

const AuthContext = React.createContext();

class AuthProvider extends React.Component {
  state = {
    isAuth: localStorage.getItem('token') ? true : false,
    token: localStorage.getItem('token') ? localStorage.getItem('token') : null,
    user: localStorage.getItem('user')
      ? JSON.parse(localStorage.getItem('user'))
      : null,
    loginError: null
  };

  constructor() {
    super();
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.isCoach = this.isCoach.bind(this);
  }

  isCoach() {
    return this.state.user.coachOfGroups.length > 0;
  }

  login(values) {
    const { email, password } = values;
    fetch('/signin', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ email, password })
    })
      .then(response =>
        response.json().then(data => ({
          status: response.status,
          token: data.token,
          user: data.user,
          message: data.message
        }))
      )
      .then(response => {
        if (response.status === 200) {
          this.setState({
            token: response.token,
            user: response.user,
            isAuth: true,
            loginError: null
          });
          localStorage.setItem('token', response.token);
          localStorage.setItem('user', JSON.stringify(response.user));
        } else {
          this.setState({ loginError: response.message });
        }
      })
      .catch(error => {
        this.setState({ loginError: error.message });
      });
  }
  logout() {
    this.setState({ isAuth: false, user: null, token: null });
    localStorage.removeItem('token');
    localStorage.removeItem('user');
  }

  render() {
    return (
      <AuthContext.Provider
        value={{
          isAuth: this.state.isAuth,
          login: this.login,
          logout: this.logout,
          loginError: this.state.loginError,
          user: this.state.user,
          isCoach: this.isCoach
        }}
      >
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

const AuthConsumer = AuthContext.Consumer;

export { AuthProvider, AuthConsumer, AuthContext };
