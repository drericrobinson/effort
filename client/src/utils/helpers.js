import { startOfWeek, endOfWeek, format } from 'date-fns';

export const weekStart = date => {
  return startOfWeek(date, { weekStartsOn: 1 });
};

export const weekEnd = date => {
  return endOfWeek(date, { weekStartsOn: 1 });
};

export const dateInWords = date => {
  return format(date, 'MMM D');
};

export const weekRangeInWords = date => {
  return (
    dateInWords(weekStart(date)) +
    String.fromCharCode(8212) +
    dateInWords(weekEnd(date))
  );
};
