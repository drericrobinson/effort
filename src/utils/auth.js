import config from '../config';
import { User } from '../resources/user/user.model';
import jwt from 'jsonwebtoken';

export const newToken = user => {
  return jwt.sign({ id: user.id }, config.secrets.jwt, {
    expiresIn: config.secrets.jwtExp
  });
};

export const verifyToken = token =>
  new Promise((resolve, reject) => {
    jwt.verify(token, config.secrets.jwt, (err, payload) => {
      if (err) return reject(err);
      resolve(payload);
    });
  });

export const signup = async (req, res) => {
  const { email, password, firstName, lastName } = req.body;

  if (!email || !password || !firstName || !lastName) {
    return res.status(400).send({
      message: 'Email, password, first name, and last name required.'
    });
  }

  try {
    const user = await User.create({ email, password, firstName, lastName });
    const token = newToken(user);

    return res.status(201).send({ token });
  } catch (e) {
    return res.status(500).send({ error: e });
  }
};

export const signin = async (req, res) => {
  const { email, password } = req.body;

  if (!email || !password) {
    return res.status(400).send({ message: 'Email and password required.' });
  }

  try {
    const user = await User.findOne({ email }).exec();
    const match = await user.checkPassword(password);

    if (!match) {
      return res.status(401).send({ message: 'Invalid password.' });
    }

    const token = newToken(user);
    return res.status(200).send({ token });
  } catch (e) {
    console.error(e);
    return res.status(401).send({ message: 'User not found.', e });
  }
};

export const authRequired = async (req, res, next) => {
  if (!req.headers.authorization) return res.status(401).end();

  const token = req.headers.authorization.split('Bearer ')[1];

  if (!token) {
    return res.status(401).end();
  }

  try {
    const payload = await verifyToken(token);
    const user = await User.findById(payload.id)
      .select('-password')
      .lean()
      .exec();

    req.user = user;
    next();
  } catch (e) {
    console.error(e);
    return res.status(401).end();
  }
};
