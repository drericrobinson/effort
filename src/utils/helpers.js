import { startOfWeek, endOfWeek, format } from 'date-fns';

export const weekStart = date => {
  return startOfWeek(date, { options: { weekStartsOn: 1 } });
};

export const weekEnd = date => {
  return endOfWeek(date, { options: { weekStartsOn: 1 } });
};

export const dateInWords = date => {
  return format(date, 'MMM D');
};

export const sendErrorMessage = (error, res) => {
  error.message === 'Unauthorized'
    ? res.status(401).json({
        error: "You don't have permission to access."
      })
    : res.status(400).end();
};
