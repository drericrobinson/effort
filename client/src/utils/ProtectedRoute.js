import React from 'react';
import { Redirect } from '@reach/router';
import { AuthConsumer } from './AuthContext';

export const ProtectedRoute = ({
  component: Component,
  altComponent: AltComponent = null,
  children,
  ...rest
}) => (
  <AuthConsumer>
    {({ isAuth }) =>
      isAuth ? (
        <Component {...rest}>{children}</Component>
      ) : AltComponent ? (
        <AltComponent {...rest}>{children}</AltComponent>
      ) : (
        <Redirect from='' to='login' noThrow />
      )
    }
  </AuthConsumer>
);
