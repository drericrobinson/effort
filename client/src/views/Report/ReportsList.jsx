import React, { Component } from 'react';
import { UikHeadline, UikWidget, UikWidgetTable, UikButton } from '../../@uik';
import { weekRangeInWords } from '../../utils/helpers';
import { withAuthContext } from '../../utils/AuthContext';
import { Link } from '@reach/router';

export class ReportsList extends Component {
  state = {
    reports: []
  };

  componentDidMount() {
    this.fetchReports();
  }

  fetchReports() {
    const groupId = this.props.user.coachOfGroups[0]._id;

    fetch(`/api/group/${groupId}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${localStorage.getItem('token')}`
      }
    })
      .then(response => response.json())
      .then(response => {
        this.setState({
          reports: response.data[0].reports.reverse()
        });
      });
  }

  render() {
    return (
      <div style={{ padding: '30px', marginLeft: '20px' }}>
        <div
          style={{
            display: 'flex',
            justifyContent: 'space-between',
            alignItems: 'center'
          }}
        >
          <UikHeadline>Reports</UikHeadline>
          <UikButton
            success
            Component={Link}
            to='/report/create'
            style={{ textDecoration: 'none' }}
          >
            New Report
          </UikButton>
        </div>
        <UikWidget>
          <UikWidgetTable>
            <thead>
              <tr>
                <th>Report</th>
                <th>Points Expected</th>
                <th>Actions</th>
              </tr>
            </thead>
            <tbody>
              {this.state.reports.length > 0 &&
                this.state.reports.map(report => {
                  return (
                    <tr key={report._id}>
                      <td>{weekRangeInWords(report.dueDate)}</td>
                      <td>{report.pointsExpected}</td>
                      <td>
                        <UikButton
                          xs
                          Component={Link}
                          to={`/group/week/${report._id}`}
                          style={{ textDecoration: 'none' }}
                        >
                          View Responses
                        </UikButton>
                      </td>
                    </tr>
                  );
                })}
            </tbody>
          </UikWidgetTable>
        </UikWidget>
      </div>
    );
  }
}

export default withAuthContext(ReportsList);
