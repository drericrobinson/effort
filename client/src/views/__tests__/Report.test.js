import React from 'react';
import { render, cleanup, waitForElement } from 'react-testing-library';
import { AuthContext } from '../../utils/AuthContext';
import ReportsList from '../Report/ReportsList';
import mongoose from 'mongoose';
import 'jest-dom/extend-expect';
import fetchMock from 'fetch-mock';

afterEach(cleanup, fetchMock.restore());

describe('<ReportList />', () => {
  it('should render correctly for coaches', async () => {
    const groupId = mongoose.Types.ObjectId();

    fetchMock.get(`/api/group/${groupId}`, {
      data: [
        {
          reports: [
            {
              _id: mongoose.Types.ObjectId(),
              dueDate: Date.now(),
              pointsExpected: 100
            }
          ]
        }
      ]
    });

    const tree = (
      <AuthContext.Provider
        value={{
          user: { coachOfGroups: [groupId] },
          isCoach: () => true
        }}
      >
        <ReportsList />
      </AuthContext.Provider>
    );

    const { getByText, container } = render(tree);
    const pointsExpected = await waitForElement(() => getByText(/100/));
    expect(container).toMatchSnapshot();
    expect(pointsExpected).toHaveTextContent('100');
    expect(getByText(/New Report/)).toBeInTheDocument();
  });
});
