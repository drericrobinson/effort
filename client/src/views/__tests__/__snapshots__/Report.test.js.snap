// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportList /> should render correctly for coaches 1`] = `
<div>
  <div
    style="padding: 30px; margin-left: 20px;"
  >
    <div
      style="display: flex; justify-content: space-between; align-items: center;"
    >
      <h2
        class="uik-headline__wrapper"
      >
        Reports
      </h2>
      <a
        class="uik-btn__base uik-btn__success"
        href="/report/create"
        style="text-decoration: none;"
      >
        <span
          class="uik-btn__content"
        >
          New Report
        </span>
      </a>
    </div>
    <div
      class="uik-widget__wrapper"
    >
      <table
        class="uik-widget-table__wrapper"
      >
        <thead>
          <tr>
            <th>
              Report
            </th>
            <th>
              Points Expected
            </th>
            <th>
              Actions
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              May 27â€”Jun 2
            </td>
            <td>
              100
            </td>
            <td>
              <a
                class="uik-btn__base uik-btn__xs"
                href="/group/week/5cf2b584c67b0387a04070e0"
                style="text-decoration: none;"
              >
                <span
                  class="uik-btn__content"
                >
                  View Responses
                </span>
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
